File: 0328-odd-even-linked-list/0328-odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {

        ListNode oddHead=null;
        ListNode oddTail=null;

        ListNode evenHead=null;
        ListNode evenTail=null;

        ListNode current=head;
        int size=1;
        while(current!=null){
            ListNode newNode = new ListNode(current.val);
            if(size%2!=0){
                if(oddHead==null){
                    oddHead=newNode;
                    oddTail=newNode;
                }
                else{
                    oddTail.next=newNode;
                    oddTail=newNode;
                }
            }
            else{
                if(evenHead==null){
                    evenHead=newNode;
                    evenTail=newNode;
                }
                else{
                    evenTail.next=newNode;
                    evenTail=newNode;
                }
            }
            current=current.next;
            size++;
        }
        ListNode newHead=new ListNode();
        if(oddHead!=null){
            newHead=oddHead;
            if(evenHead!=null){
                oddTail.next=evenHead;
            }
            else{
                oddTail.next=null;
            }
        }
        else{
            newHead=evenHead;
        }
        return newHead;
    }
}

File: 0328-odd-even-linked-list/0328-odd-even-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(head==NULL || head->next==NULL){
            return head;
        }
        ListNode* oddHead=NULL;
        ListNode* oddTail=NULL;

        ListNode* evenHead=NULL;
        ListNode* evenTail=NULL;

        ListNode* temp=head;
        int position=1;
        while(temp!=NULL){
            ListNode* newNode=new ListNode(temp->val);
            if(position%2==0){
                if(evenHead==NULL && evenTail==NULL){
                    evenHead=newNode;
                    evenTail=newNode;
                }
                else{
                    evenTail->next=newNode;
                    evenTail=newNode;
                }
            }
            else{
                if(oddHead==NULL && oddTail==NULL){
                    oddHead=newNode;
                    oddTail=newNode;
                }
                else{
                    oddTail->next=newNode;
                    oddTail=newNode;
                }
            }
            temp=temp->next;
            position++;
        }

        oddTail->next=evenHead;

        return oddHead;


    }
};